### ISSUE TITLE
Support Arrow and ArrowStream `format`

### ISSUE THREAD

**Issue by mike-luabase** [ClickHouse/clickhouse-connect#17]

**Comment by genzgd** [comment]
Thanks for opening the first issue!

Can you elaborate further on the use case here?  This sounds like an API request where the library will support direct inserts of Arrow data structures and queries that return Arrow data structures without the current conversion to ClickHouse Native for data transport.  Is that an accurate description?  Is the main goal performance improvements?  Or just to avoid the conversion of the Arrow data structure to/from native Python datatypes?

**Comment by mike-luabase** [comment]
I want to stream the Arrow data into duckdb (which is running in the browser)

https://duckdb.org/2021/10/29/duckdb-wasm.html

**Comment by genzgd** [comment]
So `someapi` would be Python?  Or do you want to do this directly from ClickHouse?

**Comment by mike-luabase** [comment]
@genzgd directly from clickhouse would be a great option, but not sure how we'd handle auth for that.

The API is written in Python

```
import requests
from requests.utils import quote

BASE_URL = f'https://{CH_USER}:{CH_PASSWORD}@{CH_HOST}'
sql = 'select 1, 2, 3 format Arrow'
sql = requests.utils.quote(sql)
d = requests.get(f'{BASE_URL}/?query={sql}')
```

**Comment by genzgd** [comment]
I will take a look next week but yeah I think direct ClickHouse Arrow format to PyArrow would be nice

**Comment by mike-luabase** [comment]
Hey @genzgd ! Any luck with this? Would it be easier to just add a `raw()` method? Then you could put the format in the query:

```
select 1, 2, 3 format Arrow
```

**Comment by mike-luabase** [comment]
Hey @genzgd ! Any luck with this? Would it be easier to just add a `raw()` method? Then you could put the format in the query:

```
select 1, 2, 3 format Arrow
```

**Comment by genzgd** [comment]
There is some of this in master that I haven't released yet because I'm waiting on some other pieces and haven't had time, but I can cut a quick release if you want a take a look.  I did anticipate the raw method. :)


```
    @abstractmethod
    def raw_query(self,
                  query: str,
                  parameters: Optional[Dict[str, Any]] = None,
                  settings: Optional[Dict[str, Any]] = None,
                  fmt: str = None) -> bytes:
        """
        Query method that simply returns the raw ClickHouse format bytes
        :param query: Query statement/format string
        :param parameters: Optional dictionary used to format the query
        :param settings: Optional dictionary of ClickHouse settings (key/string values)
        :param fmt: ClickHouse output format
        :return: bytes representing raw ClickHouse return value based on format
        """
  
```

Plus this method uses the ArrowStream format:

```
   def query_arrow(self,
                    query: str,
                    parameters: Optional[Dict[str, Any]] = None,
                    settings: Optional[Dict[str, Any]] = None,
                    use_strings: bool = True):
        """
        Query method using the ClickHouse ArrowStream format to return a PyArrow result
        :param query: Query statement/format string
        :param parameters: Optional dictionary used to format the query
        :param settings: Optional dictionary of ClickHouse settings (key/string values)
        :param use_strings:  Convert ClickHouse String type to Arrow string type (instead of binary)
        :return: Tuple of the PyArrow schema and a single record batch
        """
```

True Arrow streaming direct from ClickHouse isn't actually supported by ClickHouse server, so we'd have to do some batching or something on the client side for very large queries.  Another thing that's on my internal roadmap.

I'll cut a release with these now.

**Comment by genzgd** [comment]
Actually I was mistaken, the current release 0.1.5 has these methods.  I haven't closed this because I don't consider arrow support complete, especially without the ability to directly insert PyArrow record batches.

**Comment by steven-luabase** [comment]
@genzgd Question on the `query_arrow` method: 
In the comments where it says `:return: Tuple of the PyArrow schema and a single record batch`, does that mean the returned results are truncated and only returns the first record batch of possibly many for a given table/query? Reason I ask is I noticed that `query_arrow` returns less rows than `query_df` does for the exact same query.

**Comment by genzgd** [comment]
@steven-luabase

I just had a chance to look into the details of this and it looks like a limitation of the ClickHouse `ArrowStream` format when it is used over HTTP.  ClickHouse will return a single stream block that is limited by the `max_block_size` setting and there's no built in mechanism over HTTP to get multiple blocks with a single query (it might work in native format).

I'm looking at an alternative approach that uses the ClickHouse `Arrow` format instead of `ArrowStream`.  The downside is of course there's no "streaming" on the Python side, everything is processed as one large batch in the client.
By using a PyArrow RecordBatchFileReader I can get the results as a `PyArrow.Table` or alternatively as a tuple of the schema and a list of PyArrow.RecordBatches.   Which output is more useful in your case?

**Comment by genzgd** [comment]
Okay, the new release (0.2.0) is now available.  The query_arrow function now returns a PyArrow.Table that should not be missing data.

**Comment by steven-luabase** [comment]
Awesome, thanks so much. I'll give this a try!

**Comment by genzgd** [comment]
The most recent version (0.2.3) has a new client method arrow_insert which will insert a full PyArrow table.  If there is additional Arrow integration that can not be done with the client via the generic raw_insert or raw_query client methods feel free to open another issue.
